### PR TITLE
Add version reporting and flash bound checking

### PR DESCRIPTION
**Pull request by Jeff-Ciesielski**
This prevents the bootloader from attempting to overwrite itself, and
gives the app a way to report which bootloader version is present

Additionally, dependencies were updated to their minimum viable versions

### CODE REVIEW

**Comment by Jeff-Ciesielski** [comment]
This is to ublock/clean up RCT stuff.  MK2 is still failing, and it appears to be due to the CRC failing during the image search.  I'll do another PR soon once I figure out why that is (and will probably at some point want to move this from beta to 1.0.0)

File: host_tools/fw_update.py [hunk 1]
```diff
@@ -80,18 +102,26 @@ def scan_for_device(self):
...
         for p in ports:
             try:
                 rsp = self._send_ping(p)
                 if not rsp == None:
-                    self._log("Found bootloader on: {}".format(p))
+                    ver = self._get_version(p)
+                    if not ver:
+                        version = '<=0.1.2'
+                    else:
+                        version = '.'.join(
```
Comment (stieg) [inline host_tools/fw_update.py:113]: Who provides this version object.  Would be very nice if this code was in there instead of here where it is messy.

File: host_tools/fw_update.py [hunk 2]
```diff
@@ -80,18 +102,26 @@ def scan_for_device(self):
...
+                            [str(x) for x in
+                             (ver.major, ver.minor, ver.bugfix)])
+                    self._log("Found bootloader ver: {} on: {}".format(version, p))
                     retport = p
                     break
                 else:
                     raise Exception()
 
-            except:
+            except Exception as e:
```
Comment (stieg) [inline host_tools/fw_update.py]: Well this is not good.  Is there a more explicit exception to catch here or series of Exceptions.  I'm quite sure you don't want to catch a SIGINT here in example.

File: setup.py [hunk 3]
```diff
@@ -8,9 +9,26 @@
 def read(fname):
     return open(os.path.join(os.path.dirname(__file__), fname)).read()
 
+def extract_version():
+    with open('version.h') as inf:
+        lines = inf.readlines()
+
+    ver_map = {'MAJOR': 0,
```
Comment (stieg) [inline setup.py:16]: Ahh... we do control this object.  Can you just make a simple version object out of it with a cleaner `to_str` method

File: setup.py [hunk 4]
```diff
@@ -31,8 +49,9 @@ def read(fname):
         ]
     },
     install_requires=[
-        'ihextools >= 1.0.0',
+        'ihextools >= 1.1.0',
         'pyserial >= 2.7',
         'crcmod >= 1.7',
+        'XBVC >= 0.0.1'
```
Comment (stieg) [inline setup.py:55]: Thanks for adding this.

File: upgrade_agent/upgrade_agent_handlers.c [hunk 5]
```diff
@@ -3,28 +3,42 @@
...
 #include <upgrade_agent.h>
+#include <version.h>
 
 static struct app_info_block *last_known_info;
+extern uint32_t _flash_start;
+extern uint32_t _flash_end;
+
+static int offset_in_bootloader(uint32_t offset)
+{
+	return !!((offset < (uint32_t)&_flash_end) &&
```
Comment (stieg) [inline upgrade_agent/upgrade_agent_handlers.c]: If using `!!` here then why not use a boolean type instead?  Unless you are not using C99 or later.

File: host_tools/fw_update.py (hunk 1 above)
Comment (Jeff-Ciesielski) [inline host_tools/fw_update.py]: This is not the same object.

File: setup.py (hunk 3 above)
Comment (Jeff-Ciesielski) [inline setup.py:16]: This isn't the same object.

File: upgrade_agent/upgrade_agent_handlers.c (hunk 5 above)
Comment (Jeff-Ciesielski) [inline upgrade_agent/upgrade_agent_handlers.c]: Was just a clamp, but can switch to a bool. No logical difference.

File: host_tools/fw_update.py (hunk 2 above)
Comment (Jeff-Ciesielski) [inline host_tools/fw_update.py]: I'll expand this out to be more specific.

File: host_tools/fw_update.py (hunk 1 above)
Comment (Jeff-Ciesielski) [inline host_tools/fw_update.py:113]: This object is autogenerated.  It has a str object, but it is a lot more verbose than just printing the concatenated 'x.x.x' format.  Since the autogen system has no way to know what exactly a person wants to do with the end result objects, it's not really possible to generically have this sort of method.

File: upgrade_agent/upgrade_agent_handlers.c [hunk 6]
```diff
@@ -3,28 +3,42 @@
...
 	struct x_flash_response rsp;
 	int8_t res;
-       
-	rsp.error = ERR_SUCCESS;
 
-	res = flash_write_block(msg->data, msg->data_len, msg->offset);
+	rsp.error = ERR_SUCCESS;
 
-	if (res)
+	if (offset_in_bootloader(msg->offset)) {
```
Comment (stieg) [inline upgrade_agent/upgrade_agent_handlers.c]: You can save yourself some code and variables here if you flip the logic:

``` c
    rsp.error = ERR_FLASH_FAIL;
    if (!offset_in_bootloader(msg->offset) &&
        flash_write_block(msg->data, msg->data_len, msg->offset))
        rsp.error = ERR_SUCCESS;
```

File: MANIFEST.in [hunk 7]
```diff
@@ -0,0 +1 @@
+include version.h
```
Comment (stieg) [inline MANIFEST.in]: What is this file needed for?  Also it is missing a newline which makes git cranky

File: host_tools/fw_update.py (hunk 1 above)
Comment (stieg) [inline host_tools/fw_update.py:113]: Got it.

**Comment by stieg** [comment]
Minor nits.  Hard to know about the generated code so if bits are generated then ignore comments on them.

File: MANIFEST.in (hunk 7 above)
Comment (Jeff-Ciesielski) [inline MANIFEST.in]: setup.py parses it to pull the version number (and unfortunately does this at runtime rather than when the package is generated...).  I want to limit the location where the version number is set to once place and using a regex to grab it from a few defines is a lot simpler than autogenerating a bunch of stuff when I build.

File: host_tools/fw_update.py [hunk 8]
```diff
@@ -65,8 +65,7 @@ def _get_version(self, port):
             ep.flush()
 
             res = ep.send(ver, GET_VERSION_RESPONSE_ID)
-        except Exception as e:
-            print e
+        except:
```
Comment (stieg) [inline host_tools/fw_update.py:68]: Does this not still catch all Exceptions?  Is there any specific exception you want to catch here?

File: host_tools/fw_update.py [hunk 9]
```diff
@@ -118,10 +117,8 @@ def scan_for_device(self):
                     retport = p
                     break
                 else:
-                    raise Exception()
-
-            except Exception as e:
-                print e
+                    raise IOError()
+            except IOError:
```
Comment (stieg) [inline host_tools/fw_update.py:121]: :+1:

File: upgrade_agent/upgrade_agent_handlers.c [hunk 10]
```diff
@@ -18,19 +18,12 @@ static int offset_in_bootloader(uint32_t offset)
...
-	int8_t res;
 
-	rsp.error = ERR_SUCCESS;
-
-	if (offset_in_bootloader(msg->offset)) {
-		rsp.error = ERR_FLASH_FAIL;
-	} else {
+	rsp.error = ERR_FLASH_FAIL;
+	if (!offset_in_bootloader(msg->offset) &&
+	    !flash_write_block(msg->data, msg->data_len, msg->offset)) {
```
Comment (stieg) [inline upgrade_agent/upgrade_agent_handlers.c:24]: Extra bang leads to bad logic if I'm reading it correctly.  Should be:

``` C
if (!offset_in_bootloader(msg->offset) &&
    flash_write_block(msg->data, msg->data_len, msg->offset)) {
```

File: upgrade_agent/upgrade_agent_handlers.c [hunk 11]
```diff
@@ -3,28 +3,35 @@
...
-       
-	rsp.error = ERR_SUCCESS;
 
-	res = flash_write_block(msg->data, msg->data_len, msg->offset);
-
-	if (res)
-		rsp.error = ERR_FLASH_FAIL;
+	rsp.error = ERR_FLASH_FAIL;
+	if (!offset_in_bootloader(msg->offset) &&
+	    !flash_write_block(msg->data, msg->data_len, msg->offset)) {
```
Comment (stieg) [inline upgrade_agent/upgrade_agent_handlers.c:24]: What is the return code on `flash_write_block` when successful?  0 or non-zero?

**Comment by stieg** [comment]
One question.  Once answered, :ship:

File: upgrade_agent/upgrade_agent_handlers.c (hunk 11 above)
Comment (Jeff-Ciesielski) [inline upgrade_agent/upgrade_agent_handlers.c:24]: Unix style, 0 == success.

**Comment by stieg** [comment]
Ship it then.  Does any of this address the MK2 bootloader issues or will you handle that separately?

**Comment by Jeff-Ciesielski** [comment]
Unfortunately no, I was hoping that perhaps it was something to do with the ihextools bug, but it wasn't.  CRC check fails onboard the device.  Everything seems to flash appropriately, but it just doesn't ever find a valid image.  Will be debugging that next.
